File: MuscleHamster/Models/AppRoutingState.swift
//
//  AppRoutingState.swift
//  MuscleHamster
//
//  Manages app-wide routing state for deep links and notification taps
//  Phase 08.3: Notification Tap Routing and Today Context
//

import Foundation
import SwiftUI

// MARK: - App Routing State

/// Observable object for managing app-wide routing state
@MainActor
final class AppRoutingState: ObservableObject {

    // MARK: - Singleton

    static let shared = AppRoutingState()

    // MARK: - Published State

    /// The pending destination to route to (set by notification tap, consumed by app)
    @Published var pendingDestination: DeepLinkDestination?

    /// Current notification context to display on Home
    @Published var notificationContext: NotificationContext?

    /// Whether the app should navigate to Home tab
    @Published var shouldNavigateToHome: Bool = false

    // MARK: - Init

    private init() {}

    // MARK: - Routing Methods

    /// Handle a notification tap and set appropriate routing state
    /// - Parameters:
    ///   - notificationType: The type of notification that was tapped
    ///   - hasCheckedInToday: Whether the user has any check-in today
    ///   - currentStreak: The user's current streak count
    func handleNotificationTap(
        notificationType: NotificationType,
        hasCheckedInToday: Bool,
        currentStreak: Int
    ) {
        let context = NotificationContext(
            notificationType: notificationType,
            tappedAt: Date(),
            hasCheckedInToday: hasCheckedInToday,
            currentStreak: currentStreak
        )

        // Set the context for Home to display
        notificationContext = context

        // Set destination to home with context
        pendingDestination = .homeWithNotificationContext(context)

        // Trigger navigation to home tab
        shouldNavigateToHome = true

        print("AppRoutingState: Handling notification tap - type: \(notificationType), checkedIn: \(hasCheckedInToday)")
    }

    /// Clear the pending destination after it's been consumed
    func clearPendingDestination() {
        pendingDestination = nil
        shouldNavigateToHome = false
    }

    /// Clear the notification context after it's been displayed
    func clearNotificationContext() {
        notificationContext = nil
    }

    /// Clear all routing state
    func clearAll() {
        pendingDestination = nil
        notificationContext = nil
        shouldNavigateToHome = false
    }
}

// MARK: - Deep Link Destination

/// Possible deep link destinations in the app
enum DeepLinkDestination: Equatable {
    case home
    case homeWithNotificationContext(NotificationContext)
    case workouts
    case restDayCheckIn

    static func == (lhs: DeepLinkDestination, rhs: DeepLinkDestination) -> Bool {
        switch (lhs, rhs) {
        case (.home, .home):
            return true
        case (.homeWithNotificationContext(let lhsContext), .homeWithNotificationContext(let rhsContext)):
            return lhsContext.id == rhsContext.id
        case (.workouts, .workouts):
            return true
        case (.restDayCheckIn, .restDayCheckIn):
            return true
        default:
            return false
        }
    }
}

// MARK: - Notification Context

/// Context passed from a notification tap to the destination screen
struct NotificationContext: Identifiable, Equatable {
    let id: String
    let notificationType: NotificationType
    let tappedAt: Date
    let hasCheckedInToday: Bool
    let currentStreak: Int

    init(
        notificationType: NotificationType,
        tappedAt: Date,
        hasCheckedInToday: Bool,
        currentStreak: Int
    ) {
        self.id = UUID().uuidString
        self.notificationType = notificationType
        self.tappedAt = tappedAt
        self.hasCheckedInToday = hasCheckedInToday
        self.currentStreak = currentStreak
    }

    /// The banner type to display based on context
    var bannerType: NotificationBannerType {
        if hasCheckedInToday {
            return .streakSafe
        } else if notificationType == .streakAtRisk {
            return .actionNeeded
        } else {
            return .gentleNudge
        }
    }

    /// The message to display in the banner
    var bannerMessage: String {
        switch bannerType {
        case .streakSafe:
            if currentStreak > 0 {
                return "You're all set! Your streak is safe at \(currentStreak) day\(currentStreak == 1 ? "" : "s")."
            } else {
                return "Great job checking in today! Keep it up!"
            }
        case .actionNeeded:
            return "There's still time! Check in to keep your streak going."
        case .gentleNudge:
            return "Ready when you are! Your hamster is excited to work out."
        }
    }

    /// Whether the banner should show an action button
    var hasActionButton: Bool {
        bannerType == .actionNeeded
    }

    /// The action button title
    var actionButtonTitle: String {
        "Quick Check-in"
    }

    /// Whether the banner should auto-dismiss
    var shouldAutoDismiss: Bool {
        bannerType != .actionNeeded
    }

    /// Auto-dismiss delay in seconds
    var autoDismissDelay: TimeInterval {
        8.0
    }
}

// MARK: - Notification Banner Type

/// Types of contextual banners shown after notification tap
enum NotificationBannerType {
    case streakSafe      // User already checked in
    case actionNeeded    // Streak at risk, no check-in
    case gentleNudge     // Daily reminder, no check-in

    var icon: String {
        switch self {
        case .streakSafe:
            return "checkmark.circle.fill"
        case .actionNeeded:
            return "flame.fill"
        case .gentleNudge:
            return "pawprint.fill"
        }
    }

    var backgroundColor: Color {
        switch self {
        case .streakSafe:
            return Color.green.opacity(0.15)
        case .actionNeeded:
            return Color.orange.opacity(0.15)
        case .gentleNudge:
            return Color.purple.opacity(0.15)
        }
    }

    var iconColor: Color {
        switch self {
        case .streakSafe:
            return .green
        case .actionNeeded:
            return .orange
        case .gentleNudge:
            return .purple
        }
    }

    var accessibilityLabel: String {
        switch self {
        case .streakSafe:
            return "Streak safe notification"
        case .actionNeeded:
            return "Action needed notification"
        case .gentleNudge:
            return "Workout reminder notification"
        }
    }
}

File: MuscleHamsterExpo/src/models/AppRoutingState.js
// App Routing State Model
// Ported from Swift AppRoutingState.swift
// Phase 08.3: Notification Tap Routing and Today Context

import { NotificationType } from './NotificationPreferences';
import * as Crypto from 'expo-crypto';
import Logger from '../services/LoggerService';

// Deep Link Destinations
export const DeepLinkDestination = {
  HOME: 'home',
  HOME_WITH_CONTEXT: 'homeWithNotificationContext',
  WORKOUTS: 'workouts',
  REST_DAY_CHECK_IN: 'restDayCheckIn',
};

// Notification Banner Types
export const NotificationBannerType = {
  STREAK_SAFE: 'streakSafe',      // User already checked in
  ACTION_NEEDED: 'actionNeeded',   // Streak at risk, no check-in
  GENTLE_NUDGE: 'gentleNudge',     // Daily reminder, no check-in
};

export const NotificationBannerTypeInfo = {
  [NotificationBannerType.STREAK_SAFE]: {
    icon: 'checkmark-circle',
    backgroundColor: 'rgba(76, 175, 80, 0.15)', // green
    iconColor: '#4CAF50',
    accessibilityLabel: 'Streak safe notification',
  },
  [NotificationBannerType.ACTION_NEEDED]: {
    icon: 'flame',
    backgroundColor: 'rgba(255, 152, 0, 0.15)', // orange
    iconColor: '#FF9800',
    accessibilityLabel: 'Action needed notification',
  },
  [NotificationBannerType.GENTLE_NUDGE]: {
    icon: 'paw',
    backgroundColor: 'rgba(156, 39, 176, 0.15)', // purple
    iconColor: '#9C27B0',
    accessibilityLabel: 'Workout reminder notification',
  },
};

// Create a notification context
export const createNotificationContext = ({
  notificationType,
  tappedAt = new Date(),
  hasCheckedInToday = false,
  currentStreak = 0,
}) => {
  const id = `ctx_${Date.now()}_${Crypto.randomUUID().substring(0, 8)}`;

  // Determine banner type based on context
  let bannerType;
  if (hasCheckedInToday) {
    bannerType = NotificationBannerType.STREAK_SAFE;
  } else if (notificationType === NotificationType.STREAK_AT_RISK) {
    bannerType = NotificationBannerType.ACTION_NEEDED;
  } else {
    bannerType = NotificationBannerType.GENTLE_NUDGE;
  }

  // Generate banner message
  let bannerMessage;
  switch (bannerType) {
    case NotificationBannerType.STREAK_SAFE:
      if (currentStreak > 0) {
        bannerMessage = `You're all set! Your streak is safe at ${currentStreak} day${currentStreak === 1 ? '' : 's'}.`;
      } else {
        bannerMessage = 'Great job checking in today! Keep it up!';
      }
      break;
    case NotificationBannerType.ACTION_NEEDED:
      bannerMessage = "There's still time! Check in to keep your streak going.";
      break;
    case NotificationBannerType.GENTLE_NUDGE:
    default:
      bannerMessage = 'Ready when you are! Your hamster is excited to work out.';
      break;
  }

  const hasActionButton = bannerType === NotificationBannerType.ACTION_NEEDED;
  const actionButtonTitle = 'Quick Check-in';
  const shouldAutoDismiss = bannerType !== NotificationBannerType.ACTION_NEEDED;
  const autoDismissDelay = 8000; // 8 seconds in milliseconds

  return {
    id,
    notificationType,
    tappedAt,
    hasCheckedInToday,
    currentStreak,
    bannerType,
    bannerMessage,
    hasActionButton,
    actionButtonTitle,
    shouldAutoDismiss,
    autoDismissDelay,
  };
};

// Create initial routing state
export const createInitialRoutingState = () => ({
  pendingDestination: null,
  notificationContext: null,
  shouldNavigateToHome: false,
});

// Handle a notification tap and generate routing state updates
export const handleNotificationTap = ({
  notificationType,
  hasCheckedInToday = false,
  currentStreak = 0,
}) => {
  const context = createNotificationContext({
    notificationType,
    tappedAt: new Date(),
    hasCheckedInToday,
    currentStreak,
  });

  Logger.debug(`AppRoutingState: Handling notification tap - type: ${notificationType}, checkedIn: ${hasCheckedInToday}`);

  return {
    pendingDestination: {
      type: DeepLinkDestination.HOME_WITH_CONTEXT,
      context,
    },
    notificationContext: context,
    shouldNavigateToHome: true,
  };
};

// Parse notification identifier to determine type
export const parseNotificationType = (identifier) => {
  if (identifier.startsWith('daily_reminder')) {
    return NotificationType.DAILY_REMINDER;
  } else if (identifier.startsWith('streak_at_risk')) {
    return NotificationType.STREAK_AT_RISK;
  } else if (identifier.startsWith('friend_nudge')) {
    return NotificationType.FRIEND_NUDGE;
  }
  return null;
};

// Compare notification contexts
export const areContextsEqual = (ctx1, ctx2) => {
  if (!ctx1 && !ctx2) return true;
  if (!ctx1 || !ctx2) return false;
  return ctx1.id === ctx2.id;
};

// Compare deep link destinations
export const areDestinationsEqual = (dest1, dest2) => {
  if (!dest1 && !dest2) return true;
  if (!dest1 || !dest2) return false;
  if (dest1.type !== dest2.type) return false;

  if (dest1.type === DeepLinkDestination.HOME_WITH_CONTEXT) {
    return areContextsEqual(dest1.context, dest2.context);
  }

  return true;
};
